### PR TITLE
Core: Fix retry behavior for Jdbc Client

### LINKED ISSUE #7173: Does JDBC connector uses any retry mechanism?
### Query engine

Trino

### Question

As mentioned in this [issue](https://github.com/trinodb/trino/issues/16284), we have an aurora serverless database, and we use it as our iceberg catalog. The serverless db scales up and down during the day and can reach to 0 instances, when it happens, all connections are lost. The expected behavior for the jdbc connection in this case would be to try to reconnect to the database, since it will be the trigger for starting the db instances.

Should it be possible to implement this connection retry feature? Or is it expected, the non retry behavior?

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
Fixes #7173 

There are certain `SqlExceptions` which are currently not being retried in JdbcClient but they should be retried based on their [SqlState](https://en.wikipedia.org/wiki/SQLSTATE).

Although this implementation is really only an incremental improvement, because currently the retry behavior in `ClientPoolImpl` is hardcoded to retry only once after the initial request and without any backoff. Also currently, only a single exception can be passed in for retries which is a bit limiting.

I think there's a good amount of room for refactoring ClientPoolImpl to enable custom amount of retries and multiple exceptions similar to our `Tasks` implementation.

 Right now, the way retry can be tuned is by overriding `isConnectionException` (same as what is done in HiveClientPoolImpl). I think we can take refactoring separately since it'll be a bit more intrusive and in the interim at least we can incrementally improve the behavior.

Furthermore, the exception that is currently being retried `SqlNonTransientConnectionException` should not be retried. The client should be retrying `SqlTransientConnectionException`, so this behavior has been fixed.

cc: @jackye1995 @singhpk234 @nastra @szehon-ho

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java [hunk 1]
```diff
@@ -21,17 +21,22 @@
...
 import org.apache.iceberg.ClientPoolImpl;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
 
 public class JdbcClientPool extends ClientPoolImpl<Connection, SQLException> {
 
   private final String dbUrl;
   private final Map<String, String> properties;
 
+  private static final Set<String> RETRYABLE_CONNECTION_SQL_STATES =
+      ImmutableSet.of("08000", "08003", "08006", "08004", "08007");
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java:38]: I should probably link what these SqlStates are in a code comment.

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java:38]: +1

**Review by jackye1995 — COMMENTED** [review]
looks good, would be great to add comments about the meaning of error codes, not sure if there are any library to directly import these values. Also can we add some tests?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java (hunk 1 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java:38]: Any reason for including 
- 08004 As per doc look like it's due to sql rejecting the connection. refering this doc of pg-jdbc driver mentioned in description above [doc](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/util/PSQLState.java#LL31C1-L36C32) : 
```  /**
   * The server rejected our connection attempt. Usually an authentication failure, but could be a
   * configuration error like asking for a SSL connection with a server that wasn't built with SSL
   * support.
   */
  CONNECTION_REJECTED("08004"), 
```

**Review by singhpk234 — COMMENTED** [review]
+1 for the change and making client pool, linking previous discussions on the same as well https://github.com/apache/iceberg/pull/3099

**Comment by cccs-eric** [comment]
Me and my co-worker just raised in Iceberg General Slack the same issue that this PR is fixing.  Any timeline on when this could be merged in?  And also, will this fix work for when a Postgresql database has closed an idle connection (so a reconnect will be done)?

**Comment by cccs-jc** [comment]
@amogh-jahagirdar,  Is there any progress on making the error codes configurable. This would avoid forks just to add code specific to for example posgresql.

**Comment by jean-humann** [comment]
@nastra @szehon-ho Is there anything we can do to push this, please ?
Currently, we are forking Iceberg internally just to add this part to use Iceberg Rest Catalog with JDBC to workout properly.

**Comment by amogh-jahagirdar** [comment]
Hey all @jean-humann @cccs-br @cccs-eric @cccs-jc sorry I forgot to follow up on this PR! I'm taking a look now to see how to carry this forward. It does seem like since there are codes which will not be part of the standard spec, but that still need to be retried , that we need a configurable list (since we wouldn't want to maintain such a list in Iceberg)

**Comment by amogh-jahagirdar** [comment]
@jean-humann @cccs-br @cccs-eric @cccs-jc  https://github.com/apache/iceberg/pull/10140 this is a draft PR for having a configurable set of retry codes. I'll need to write some tests for it but just a heads up since I don't think I'll be updating this branch, and would be closing this PR if there's consensus from others that having a user provided error code list is a good idea to solve the problems you all are facing! I think it is the most practical option at the moment until we separate catalogs from the reference implementation in the future https://www.youtube.com/watch?v=uAQVGd5zV4I&t=1323s

**Comment by jfuerth** [comment]
My team hit this issue specifically with PostgreSQL today. Would love it if we could at least get something in that works for Postgres in particular, since that seems to be the most common thing that brings people here.

Please let me know how we can help move this forward!

**Pull request closed without merging**